File: src/lexer.cpp
#include "lexer.h"

#include <cctype>
#include <stdexcept>

Lexer::Lexer(std::string source) : source_(std::move(source)) {}

std::vector<Token> Lexer::tokenize() {
    while (!isAtEnd()) {
        start_ = current_;
        scanToken();
    }

    tokens_.push_back({TokenType::EndOfFile, "", line_});
    return tokens_;
}

bool Lexer::isAtEnd() const {
    return current_ >= source_.size();
}

char Lexer::advance() {
    return source_[current_++];
}

char Lexer::peek() const {
    if (isAtEnd()) {
        return '\0';
    }
    return source_[current_];
}

char Lexer::peekNext() const {
    if (current_ + 1 >= source_.size()) {
        return '\0';
    }
    return source_[current_ + 1];
}

bool Lexer::match(char expected) {
    if (isAtEnd() || source_[current_] != expected) {
        return false;
    }
    ++current_;
    return true;
}

void Lexer::addToken(TokenType type, const std::string& lexeme) {
    tokens_.push_back({type, lexeme, line_});
}

void Lexer::scanToken() {
    char c = advance();
    switch (c) {
    case '(':
        addToken(TokenType::LeftParen, "(");
        break;
    case ')':
        addToken(TokenType::RightParen, ")");
        break;
    case '/':
        if (match('/')) {
            while (!isAtEnd() && peek() != '\n') {
                advance();
            }
        } else {
            throw std::runtime_error("Use English keywords like 'over' instead of '/' at line " + std::to_string(line_));
        }
        break;
    case ';':
        addToken(TokenType::Semicolon, ";");
        break;
    case ' ':
    case '\r':
    case '\t':
        break;
    case '\n':
        addToken(TokenType::Newline, "\\n");
        ++line_;
        break;
    default:
        if (std::isdigit(static_cast<unsigned char>(c))) {
            scanNumber();
        } else if (std::isalpha(static_cast<unsigned char>(c)) || c == '_') {
            scanIdentifier();
        } else {
            throw std::runtime_error("Unexpected character '" + std::string(1, c) + "' at line " + std::to_string(line_));
        }
        break;
    }
}

void Lexer::scanNumber() {
    while (std::isdigit(static_cast<unsigned char>(peek()))) {
        advance();
    }

    addToken(TokenType::Number, source_.substr(start_, current_ - start_));
}

void Lexer::scanIdentifier() {
    while (std::isalnum(static_cast<unsigned char>(peek())) || peek() == '_') {
        advance();
    }

    const std::string text = source_.substr(start_, current_ - start_);
    if (text == "let") {
        addToken(TokenType::Let, text);
    } else if (text == "set") {
        addToken(TokenType::Set, text);
    } else if (text == "be") {
        addToken(TokenType::Be, text);
    } else if (text == "to") {
        addToken(TokenType::To, text);
    } else if (text == "print") {
        addToken(TokenType::Print, text);
    } else if (text == "if") {
        addToken(TokenType::If, text);
    } else if (text == "else") {
        addToken(TokenType::Else, text);
    } else if (text == "while") {
        addToken(TokenType::While, text);
    } else if (text == "do") {
        addToken(TokenType::Do, text);
    } else if (text == "end") {
        addToken(TokenType::End, text);
    } else if (text == "is") {
        addToken(TokenType::Is, text);
    } else if (text == "not") {
        addToken(TokenType::Not, text);
    } else if (text == "less") {
        addToken(TokenType::Less, text);
    } else if (text == "than") {
        addToken(TokenType::Than, text);
    } else if (text == "greater") {
        addToken(TokenType::Greater, text);
    } else if (text == "or") {
        addToken(TokenType::Or, text);
    } else if (text == "equal") {
        addToken(TokenType::Equal, text);
    } else if (text == "plus") {
        addToken(TokenType::Plus, text);
    } else if (text == "minus") {
        addToken(TokenType::Minus, text);
    } else if (text == "times") {
        addToken(TokenType::Times, text);
    } else if (text == "over") {
        addToken(TokenType::Over, text);
    } else if (text == "negative") {
        addToken(TokenType::Negative, text);
    } else {
        addToken(TokenType::Identifier, text);
    }
}

File: rust/src/lexer.rs
use crate::token::{Token, TokenType};

pub struct Lexer {
    source: Vec<char>,
    tokens: Vec<Token>,
    start: usize,
    current: usize,
    line: usize,
}

impl Lexer {
    pub fn new(source: String) -> Self {
        Self {
            source: source.chars().collect(),
            tokens: Vec::new(),
            start: 0,
            current: 0,
            line: 1,
        }
    }

    pub fn tokenize(mut self) -> Result<Vec<Token>, String> {
        while !self.is_at_end() {
            self.start = self.current;
            self.scan_token()?;
        }

        self.tokens.push(Token {
            token_type: TokenType::EndOfFile,
            lexeme: String::new(),
            line: self.line,
        });

        Ok(self.tokens)
    }

    fn is_at_end(&self) -> bool {
        self.current >= self.source.len()
    }

    fn advance(&mut self) -> char {
        let ch = self.source[self.current];
        self.current += 1;
        ch
    }

    fn peek(&self) -> char {
        if self.is_at_end() {
            '\0'
        } else {
            self.source[self.current]
        }
    }

    fn match_char(&mut self, expected: char) -> bool {
        if self.is_at_end() || self.source[self.current] != expected {
            return false;
        }
        self.current += 1;
        true
    }

    fn add_token(&mut self, token_type: TokenType, lexeme: String) {
        self.tokens.push(Token {
            token_type,
            lexeme,
            line: self.line,
        });
    }

    fn scan_token(&mut self) -> Result<(), String> {
        let ch = self.advance();
        match ch {
            '(' => self.add_token(TokenType::LeftParen, "(".to_string()),
            ')' => self.add_token(TokenType::RightParen, ")".to_string()),
            '/' => {
                if self.match_char('/') {
                    while !self.is_at_end() && self.peek() != '\n' {
                        self.advance();
                    }
                } else {
                    return Err(format!(
                        "Use English keywords like 'over' instead of '/' at line {}",
                        self.line
                    ));
                }
            }
            ';' => self.add_token(TokenType::Semicolon, ";".to_string()),
            ' ' | '\r' | '\t' => {}
            '\n' => {
                self.add_token(TokenType::Newline, "\\n".to_string());
                self.line += 1;
            }
            _ if ch.is_ascii_digit() => self.scan_number(),
            _ if ch.is_ascii_alphabetic() || ch == '_' => self.scan_identifier(),
            _ => {
                return Err(format!(
                    "Unexpected character '{}' at line {}",
                    ch, self.line
                ))
            }
        }

        Ok(())
    }

    fn scan_number(&mut self) {
        while self.peek().is_ascii_digit() {
            self.advance();
        }

        let lexeme: String = self.source[self.start..self.current].iter().collect();
        self.add_token(TokenType::Number, lexeme);
    }

    fn scan_identifier(&mut self) {
        while self.peek().is_ascii_alphanumeric() || self.peek() == '_' {
            self.advance();
        }

        let lexeme: String = self.source[self.start..self.current].iter().collect();
        let token_type = match lexeme.as_str() {
            "let" => TokenType::Let,
            "set" => TokenType::Set,
            "be" => TokenType::Be,
            "to" => TokenType::To,
            "print" => TokenType::Print,
            "if" => TokenType::If,
            "else" => TokenType::Else,
            "while" => TokenType::While,
            "do" => TokenType::Do,
            "end" => TokenType::End,
            "is" => TokenType::Is,
            "not" => TokenType::Not,
            "less" => TokenType::Less,
            "than" => TokenType::Than,
            "greater" => TokenType::Greater,
            "or" => TokenType::Or,
            "equal" => TokenType::Equal,
            "plus" => TokenType::Plus,
            "minus" => TokenType::Minus,
            "times" => TokenType::Times,
            "over" => TokenType::Over,
            "negative" => TokenType::Negative,
            _ => TokenType::Identifier,
        };

        self.add_token(token_type, lexeme);
    }
}
